include/seastar/coroutine/async_generator.hh

```diff
@@ -0,0 +1,434 @@
+    bool finished() const noexcept {
+        return _value == nullptr;
+    }
+
+    void rethrow_if_unhandled_exception() {
+        if (_exception) {
+            std::rethrow_exception(std::move(_exception));
+        }
+    }
+
+    void run_and_dispose() noexcept final {
+        using handle_type = std::coroutine_handle<async_generator_promise_base>;
+        handle_type::from_promise(*this).resume();
+    }
+
+    seastar::task* waiting_task() noexcept final {
+        return _waiting_task;
+    }
+
+    class element_awaiter {
+        std::remove_cvref_t<Yielded> _value;
+        constexpr bool await_ready() const noexcept {
+            return false;
+        }
+        template <typename Promise>
+        constexpr void await_suspend(std::coroutine_handle<Promise> producer) noexcept {
+            auto& current = producer.promise();
+            producer._value = std::addressof(_value);
+        }
+        constexpr void await_resume() const noexcept {}
+    };
+
+private:
+    friend class next_awaiter<Yielded>;
+};
+
+template <typename Yielded>
+class next_awaiter {
+protected:
+    async_generator_promise_base<Yielded>* _promise = nullptr;
+    std::coroutine_handle<> _producer = nullptr;
+
+    explicit next_awaiter(std::nullptr_t) noexcept {}
+    next_awaiter(async_generator_promise_base<Yielded>& promise,
+                 std::coroutine_handle<> producer) noexcept
+        : _promise{std::addressof(promise)}
+        , _producer{producer} {}
+
+public:
+    bool await_ready() const noexcept {
+        return false;
+    }
+
+    template <typename Promise>
+    std::coroutine_handle<> await_suspend(std::coroutine_handle<Promise> consumer) noexcept {
+        _promise->_consumer = consumer;
+        return _producer;
+    }
+};
+
+} // namespace internal
+
+/// async_generator represents a view modeling std::ranges::input_range,
+/// and has move-only iterators.
+///
+/// async_generator has 2 template parameters:
+///
+/// - Ref
+/// - Value
+///
+/// From Ref and Value, we derive types:
+/// - value_type: a cv-unqualified object type that specifies the value type of
+///   the generator's range and iterators
+/// - reference_Type: the reference type of the generator's range and iterators
+/// - yielded_type: the type of the parameter to the primary overload of \c
+///   yield_value in the generator's associated promise type
+///
+/// Under the most circumstances, only the first parameter is specified: like
+/// \c generator<meow>. The resulting generator:
+/// - has a value type of \c remove_cvref_t<meow>
+/// - has a reference type of \c meow, if it is a reference type, or \c meow&&
+///   otherwise
+/// - the operand of \c co_yield in the body of the generator should be
+///   convertible to \c meow, if it is a reference type, otherwise the operand
+///   type should be <tt>const meow&</tt>
+///
+/// Consider following code snippet:
+/// \code
+/// async_generator<const std::string&> send_query(std::string query) {
```

Comment:
why isn't this expressed as `async_generator<std::string>`?

Reply:
Intuitively, I'd rather specify `Value` and let the `Ref` type be implicit derived if omitted.

Reply:
i intended to be more compliant to the standard. see https://en.cppreference.com/w/cpp/coroutine/generator, or the original proposal, see https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2502r2.pdf .